### PR TITLE
[WebSocketServer] Allow to extend the original `WebSocket` class

### PR DESCRIPTION
**Pull request by vansergen**
@lpinca What do you think about adding this feature?

### CODE REVIEW

**Comment by lpinca** [comment]
I'm not very happy with it. Are there big advantages over monkey patching the `WebSocket` class directly?

```js
const WebSocket = require('ws');

Object.defineProperty(WebSocket.prototype, 'custom_field', {
  get() {
    return '123';
  }
});
```

**Comment by vansergen** [comment]
> Are there big advantages over monkey patching the WebSocket class directly?

I do believe so. It would give more flexibility in terms of accessing the
original methods while overwriting them

```javascript
class CustomWebSocket extends WebSocket {
  #sent;
  #created_at;
  constructor(...params) {
    super(...params);
    this.#sent = 0;
    this.#created_at = new Date();
  }
  get sent() {
    return this.#sent;
  }
  get created_at() {
    return new Date(this.#created_at);
  }
  send(...params) {
    this.#sent++;
    super.send(...params);
  }
}
```

in the same manner as the `node:http` module does

```javascript
import { IncomingMessage, ServerResponse, createServer } from 'node:http';
class CustomIncomingMessage extends IncomingMessage {}
class CustomServerResponse extends ServerResponse {}
createServer({
  IncomingMessage: CustomIncomingMessage,
  ServerResponse: CustomServerResponse
}).on('request', (request, response) => {
  console.log(request instanceof CustomIncomingMessage);
  console.log(response instanceof CustomServerResponse);
});
```

In this way one can modify `WebSocket` classes per server (without modifying the original class) 

```javascript
import { createServer } from 'http';
import { parse } from 'url';
import { WebSocketServer, WebSocket } from 'ws';

class WebSocket1 extends WebSocket {}
class WebSocket2 extends WebSocket {}

const server = createServer();
const wss1 = new WebSocketServer({ WebSocket: WebSocket1, noServer: true });
const wss2 = new WebSocketServer({ WebSocket: WebSocket2, noServer: true });

wss1.on('connection', function connection(ws) {
  // console.log(ws instanceof WebSocket1)
});
wss2.on('connection', function connection(ws) {
  // console.log(ws instanceof WebSocket2)
});

server.on('upgrade', function upgrade(request, socket, head) {
  const { pathname } = parse(request.url);
  if (pathname === '/foo') {
    wss1.handleUpgrade(request, socket, head, function done(ws) {
      wss1.emit('connection', ws, request);
    });
  } else if (pathname === '/bar') {
    wss2.handleUpgrade(request, socket, head, function done(ws) {
      wss2.emit('connection', ws, request);
    });
  } else {
    socket.destroy();
  }
});
server.listen(8080);
```

**Comment by lpinca** [comment]
Hmm, ok. It might break if the class does not extend `WebSocket` and I do not like to keep adding options but it is reasonable.

File: test/websocket-server.test.js [hunk 1]
```diff
@@ -510,6 +510,41 @@ describe('WebSocketServer', () => {
...
+
+        const ws = new WebSocket(`ws://localhost:${server.address().port}`);
+
+        ws.on('open', () => {
+          ws.on('close', () => {
+            server.close(done);
+          });
+        });
+      });
+    });
```
Comment (lpinca) [inline test/websocket-server.test.js:547]:
```suggestion
    it('honors the `WebSocket` option', (done) => {
      class CustomWebSocket extends WebSocket.WebSocket {
        get foo() {
          return 'foo';
        }
      }

      const wss = new WebSocket.Server(
        {
          port: 0,
          WebSocket: CustomWebSocket
        },
        () => {
          const ws = new WebSocket(`ws://localhost:${wss.address().port}`);

          ws.on('open', ws.close);
        }
      );

      wss.on('connection', (ws) => {
        assert.ok(ws instanceof CustomWebSocket);
        assert.strictEqual(ws.foo, 'foo');
        wss.close(done);
      });
    });
```

Also, please move it under the "options" section.

File: doc/ws.md [hunk 2]
```diff
@@ -84,6 +84,8 @@ This class represents a WebSocket server. It extends the `EventEmitter`.
   - `verifyClient` {Function} A function which can be used to validate incoming
     connections. See description below. (Usage is discouraged: see
     [Issue #337](https://github.com/websockets/ws/issues/377#issuecomment-462152231))
+  - `WebSocket` {Function} Specifies the `WebSocket` class to be used. It must
+    be extended from the original `WebSocket`. _Default_: `WebSocket`.
```
Comment (lpinca) [inline doc/ws.md:88]:
```suggestion
    be extended from the original `WebSocket`. Defaults to `WebSocket`.
```

For consistency

File: test/websocket-server.test.js [hunk 3]
```diff
@@ -510,6 +510,32 @@ describe('WebSocketServer', () => {
         });
       });
     });
+
+    it('honors the `WebSocket` option', (done) => {
```
Comment (lpinca) [inline test/websocket-server.test.js:514]: Please move the test under the "options" section (line 92).

Comment (vansergen) [reply test/websocket-server.test.js]: https://github.com/websockets/ws/pull/2007/commits/50226c8a74cf99ebfae56006ccb4a09e0e4ff5f7

**Review by lpinca — APPROVED** [review]

File: lib/websocket-server.js [hunk 4]
```diff
@@ -356,7 +359,7 @@ class WebSocketServer extends EventEmitter {
       `Sec-WebSocket-Accept: ${digest}`
     ];
 
-    const ws = new WebSocket(null);
+    const ws = new this.options.WebSocket(null);
```
[the remark this answers is not in the document]
Comment (vansergen) [reply lib/websocket-server.js:362]: I got the following results (using `hrtime.bigint()`):

```javascript
const ws = new WebSocket(null); // Average `9161n` nanoseconds
const ws = new this.options.WebSocket(null); // Average `13524n` nanoseconds
```

[the remark this answers is not in the document]
Comment (vansergen) [reply lib/websocket-server.js:362]: It seems that introducing a new variable
```js 
this._WebSocket = options.WebSocket
``` 
and accessing it with
```javascript
const ws = new this._WebSocket(null);
```
would not give such an impact

[the remark this answers is not in the document]
Comment (lpinca) [reply lib/websocket-server.js:362]: I think this is acceptable, we can eventually add the new property in future.

**Comment by lpinca** [comment]
Thank you.

**Pull request merged**